crates/sncast/src/lib.rs

```diff
@@ -267,22 +267,80 @@ pub fn get_block_id(value: &str) -> Result<BlockId> {
     }
 }
 
+#[derive(Debug)]
+pub struct ErrorData {
+    pub data: String,
+}
+
+impl ErrorData {
+    #[must_use]
+    pub fn new(data: String) -> Self {
+        ErrorData { data }
+    }
+}
+
+impl From<ContractErrorData> for ErrorData {
+    fn from(value: ContractErrorData) -> Self {
+        ErrorData {
+            data: value.revert_error,
+        }
+    }
+}
+
+#[derive(Debug)]
+pub enum TransactionError {
+    Rejected,
+    Reverted(ErrorData),
+}
+
+#[derive(Debug)]
+pub enum WaitForTransactionError {
+    TransactionError(TransactionError),
+    Other(Error),
```

Comment:
My idea was to divide these into `may happen error` and `should not happen error`

Reply:
Other -> Unexpected ?

Reply:
Would make sense, but I can see this is also used for flags verification for timeouts, which is not strictly unexpected. One way to appraoch this would be to move this verification to clap (not sure if possible), then `unexpected` would ba a good name.

Reply:
I'm thinking that we should verify if these flags are correct at the moment we receive them. We can introduce a struct, smth like `WaitForTxRetries` which can only hold valid parameters. The validation may happen in the struct `new` function

Reply:
To avoid mixing stuff up I created a separate PR for this https://github.com/foundry-rs/starknet-foundry/pull/1720

Reply:
yep

Reply:
cool. let's then verify where it makes sense to use this error, and as it is unexpected we should probably panic in those places, right?

Reply:
OK, so after introducing `Err(WaitForTransactionError::TimedOut)`, the only line using Other is
```
Err(err) => return Err(WaitForTransactionError::Other(err.into())),
```

Reply:
`err` on the left side is ProviderError

Reply:
I guess those will be some network errors

Reply:
I think we can panic there, unless it's `ProviderError::RateLimited`. In this case we can just `sleep`, maybe put a msg for the user to stdout

Reply:
yup, makes sense to panic imo, and depending on the chained error messages that will be thrown here, we then could either leave this as is or change Other to just NetworkError/ProviderError

Reply:
If we panic here, there will be no need for `Other` option in `WaitForTransactionError`

Reply:
Finally, I changed `WaitForTransactionError::Other` to `WaitForTransactionError::ProviderError`, since network error is not a bug, so it's better to wrap it in an error instead of panicking. WHYT?

Reply:
Perhaps, as we rather tend to panic when there are sncast unexpected issues, but I wonder if it may make sense to panic for provider error as well, as this is not recoverable and generally something rather unexpected. So imo panic is completely fine, but if you'd like to wrap to Err more I won't oppose

Reply:
I guess panics should be used to signal a bug in sncast, and network error is something out of sncast area of responsibility